### PR TITLE
Auto partition pruning when encounters the delete statement

### LINKED ISSUE #3137: Restrict large deletes
## Enhancement
Now that the delete operation is freed from the partition restriction, users can easily send a relatively large Delete query. 
The relatively large number means that the number of Tablets marked at one time is rather large.
If the user misuses it, the delete will eventually be stuck anyway. 
So we might as well restrict not sending delete queries when we think it is too big.

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #3137

## Problem Summary(Required) ：

This is implemented using the old optimizer, which may not be perfect when pruning. A better way to use the new optimizer, but the implementation may be more complicated.

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java [hunk 1]
```diff
@@ -374,6 +402,74 @@ public void process(DeleteStmt stmt) throws DdlException, QueryStateException {
...
+                        filter.setLowerBound(literalExpr, true);
+                        break;
+                    case GT:
+                        filter.setLowerBound(literalExpr, false);
+                        break;
+                    default:
+                        break;
+                }
+                columnFilters.put(slotRef.getColumnName(), filter);
+            } else if (condition instanceof InPredicate) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java:451]: Should check and filter `not in predicate`.

File: fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java [hunk 2]
```diff
@@ -374,6 +402,74 @@ public void process(DeleteStmt stmt) throws DdlException, QueryStateException {
...
+            throws DdlException, AnalysisException {
+        Map<String, PartitionColumnFilter> columnFilters = Maps.newHashMap();
+        List<Predicate> deleteConditions = stmt.getDeleteConditions();
+        for (Predicate condition : deleteConditions) {
+            SlotRef slotRef = (SlotRef) condition.getChild(0);
+            String columnName = slotRef.getColumnName();
+            Map<String, Column> nameToColumn = Maps.newTreeMap(String.CASE_INSENSITIVE_ORDER);
+            for (Column column : table.getBaseSchema()) {
+                nameToColumn.put(column.getName(), column);
+            }
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java:415]: Build the map outside of the for loop.

File: fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java [hunk 3]
```diff
@@ -172,7 +179,28 @@ public void process(DeleteStmt stmt) throws DdlException, QueryStateException {
...
+                        if (selectedPartitionIds == null) {
+                            partitionNames.addAll(olapTable.getPartitionNames());
+                        } else {
+                            Set<String> candidatePartitionNames = olapTable.getPartitionNames();
+                            for (String candidatePartitionName : candidatePartitionNames) {
+                                Partition partition = olapTable.getPartition(candidatePartitionName);
+                                if (selectedPartitionIds.contains(partition.getId())) {
+                                    partitionNames.add(candidatePartitionName);
+                                }
+                            }
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java:199]:
```
for (Long partitionId : selectedPartitionIds) {
    Partition partition = olapTable.getPartition(partitionId);
    partitionNames.add(partition.getName());
}
```

**Review by wyb — DISMISSED** [review]

File: fe/fe-core/src/test/java/com/starrocks/load/DeletePruneTest.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
+package com.starrocks.load;
```
Comment (chaoyli) [inline fe/fe-core/src/test/java/com/starrocks/load/DeletePruneTest.java:1]: Add license in the header.

File: fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java [hunk 5]
```diff
@@ -374,6 +386,110 @@ public void process(DeleteStmt stmt) throws DdlException, QueryStateException {
...
+                        filter.setUpperBound(literalExpr, true);
+                        filter.lowerBoundInclusive = true;
+                        break;
+                    case LT:
+                        filter.setUpperBound(literalExpr, false);
+                        filter.lowerBoundInclusive = true;
+                        break;
+                    case GE:
+                        filter.setLowerBound(literalExpr, true);
+                        break;
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java:461]: Does not filter.upperBoundInclusive?

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java [hunk 6]
```diff
@@ -374,6 +386,110 @@ public void process(DeleteStmt stmt) throws DdlException, QueryStateException {
...
+                        filter.setLowerBound(literalExpr, true);
+                        filter.setUpperBound(literalExpr, true);
+                        break;
+                    case LE:
+                        filter.setUpperBound(literalExpr, true);
+                        filter.lowerBoundInclusive = true;
+                        break;
+                    case LT:
+                        filter.setUpperBound(literalExpr, false);
+                        filter.lowerBoundInclusive = true;
```
[the remark this answers is not in the document]
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/load/DeleteHandler.java:457]: This just copy ColumnFilterConverter.java 107~130. in fact lowerBoundInclusive may not be set.

**Review by chaoyli — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**